### PR TITLE
Add projection pushdown support for DISTINCT ON and ORDER BY

### PR DESCRIPTION
**Pull request by corwinjoy**
This is related to https://github.com/duckdb/duckdb/issues/10214. The underlying issue is that projection pushdown is not supported for DISTINCT ON and ORDER BY.

Here I am hoping to improve duckdb by adding projection pushdown support for these cases, but I could use a little help.

DISTINCT ON seems easy to fix, we just need to enable a check for it as far as I can tell.
ORDER BY is harder. I have a change, below, where I enable projection pushdown for order by which works for most cases. But, it breaks down when we have constant expression like '1', true, etc. that are not used at a higher level. The optimizer tries to remove these, but then lower level bindings still seem to want the bindings. I could use some help in how to handle this.

The code in the folder projection_pushdown is for illustration and debugging since I wasn't certain how to trace through the unit tests. It will be removed in the final PR.

### CODE REVIEW

File: src/optimizer/remove_unused_columns.cpp [hunk 1]
```diff
@@ -176,10 +177,22 @@ void RemoveUnusedColumns::VisitOperator(LogicalOperator &op) {
...
+				remove.VisitOperatorExpressions(op);
+				remove.column_references.insert(column_references.begin(), column_references.end()); // add parent references
+				remove.VisitOperator(*child);
+			}
+		} else {
+			for (auto &child : op.children) {
+				RemoveUnusedColumns remove(binder, context, true);
+				remove.VisitOperator(*child);
+			}
 		}
```
Comment (corwinjoy) [inline src/optimizer/remove_unused_columns.cpp:196]: The above code change for ORDER BY projection pushdown is where I could use some help. It breaks down for constant columns. I don't fully understand the optimizer code and could use some help in how to handle this case. For example:
```
# name: test/issues/rigger/test_10087.test
# description: SQLancer bug that detected an issue with the column lifetime analyzer
# group: [rigger]

statement ok
PRAGMA enable_verification

statement ok
CREATE TABLE t0(c1 INT);

statement ok
INSERT INTO t0(c1) VALUES (1);

statement ok
CREATE VIEW v0(c0, c1, c2) AS SELECT '1', true, t0.c1 FROM t0 ORDER BY -1-2 LIMIT 2;

statement ok
SELECT v0.c2 FROM v0 WHERE (NOT (v0.c1 IS NOT NULL));
```

This gives the error:
```
INTERNAL Error: Failed to bind column reference "" [1.0] (bindings: {#[1.1], #[1.2]})
```
Basically, the optimizer is trying to skip the expressions `1` and true as unused by the higher level select but the engine still seems to want to execute them.

Comment (corwinjoy) [reply src/optimizer/remove_unused_columns.cpp:197]: In the general case, this code seems to work and correctly applies projection pushdown. So, it is this edge case that I am hoping to resolve.

**Comment by Tmonster** [comment]
Hi @corwinjoy, thanks for the PR. I believe the issue has already been fixed via https://github.com/duckdb/duckdb/pull/11006. 
If you believe this is not the case, can you provide us with a query and it's `EXPLAIN` output that shows a projection is not being pushed down?

**Comment by corwinjoy** [comment]
Hi @Tmonster, thanks so much for pointing this out and thanks for the awesome PR! Was very interesting to read how you did it and am glad to report that this does indeed seem to fix the issue nicely! Anyway, much appreciate and will post to the main issue that this fixes the problem. One, possibly naive, question. For the branch duckdb:feature is this what is scheduled for version 1.1.0? Anyway, thanks again for the PR and pointing this out!

**Comment by Tmonster** [comment]
Hi @corwinjoy,

No problem, thank you for closing the issue 👍 . To answer your questions, yes, the plan is that all functionality in the feature branch will be in the eventual v1.1.0 release.

**Pull request closed without merging**